applications/ChimeraApplication/custom_utilities/hole_cutting_utility.cpp

```diff
@@ -0,0 +1,438 @@
+    // node ids representing that face The given set of node ids may have a
+    // different node order
+    hashmap_vec ordered_skin_face_nodes_map;
+
+// Fill map that gives original node order for set of nodes
+#pragma omp parallel for
+    for (int i_e = 0; i_e < num_elements; ++i_e) {
+        auto i_element = elements_begin + i_e;
+        Element::GeometryType::GeometriesArrayType faces;
+        if (TDim == 2)
+            faces = i_element->GetGeometry().GenerateEdges();
+        else if (TDim == 3)
+            faces = i_element->GetGeometry().GenerateFaces();
+
+        for (IndexType i_face = 0; i_face < faces.size(); i_face++) {
+            // Create vector that stores all node is of current i_face
+            vector<IndexType> ids(faces[i_face].size());
+            vector<IndexType> unsorted_ids(faces[i_face].size());
+
+            // Store node ids
+            for (IndexType i = 0; i < faces[i_face].size(); i++) {
+                ids[i] = faces[i_face][i].Id();
+                unsorted_ids[i] = faces[i_face][i].Id();
+            }
+
+            //*** THE ARRAY OF IDS MUST BE ORDERED!!! ***
+            std::sort(ids.begin(), ids.end());
+#pragma omp critical
+            {
+                if (n_faces_map[ids] == 1)
+                    ordered_skin_face_nodes_map[ids] = unsorted_ids;
+            }
+        }
+    }
+    // First assign to skin model part all nodes from original model_part,
+    // unnecessary nodes will be removed later
+    IndexType id_condition = 1;
+
+    // Add skin faces as triangles to skin-model-part (loop over all node sets)
+    std::vector<IndexType> vector_of_node_ids;
+    for (typename hashmap::const_iterator it = n_faces_map.begin();
+         it != n_faces_map.end(); ++it) {
+        // If given node set represents face that is not overlapping with a face
+        // of
+        // another element, add it as skin element
+        if (it->second == 1) {
+            // If skin edge is a triangle store triangle in with its original
+            // orientation in new skin model part
+            if (it->first.size() == 2) {
+                // Getting original order is important to properly reproduce
+                // skin edge
+                // including its normal orientation
+                vector<IndexType> original_nodes_order =
+                    ordered_skin_face_nodes_map[it->first];
+
+                Node<3>::Pointer pnode1 =
+                    rVolumeModelPart.Nodes()(original_nodes_order[0]);
+                Node<3>::Pointer pnode2 =
+                    rVolumeModelPart.Nodes()(original_nodes_order[1]);
+
+                // Storing the node ids list
+                vector_of_node_ids.push_back(original_nodes_order[0]);
+                vector_of_node_ids.push_back(original_nodes_order[1]);
+
+                Properties::Pointer properties =
+                    rExtractedBoundaryModelPart.rProperties()(0);
+                Condition const& rReferenceLineCondition =
```

Comment:
ouchouchouchouchouch
Inside the loop :sweat_smile:

Reply:
Ouch !